core/types/schema.go

```diff
@@ -804,3 +742,228 @@ func (m *Modifier) Clean() error {
+			return t.Clean()
+		}
+	}
+
+	if p.Table != nil {
+		for _, col := range p.Table {
+			err := col.Clean()
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// it is possible that both are nil
+	return nil
+}
+
+// ProcedureParameter is a parameter in a procedure.
+type ProcedureParameter struct {
+	// Name is the name of the parameter.
+	// It should always be lower case.
+	Name string `json:"name"`
+	// Type is the type of the parameter.
+	Type *DataType `json:"type"`
+}
+
+func (c *ProcedureParameter) Clean() error {
+	return errors.Join(
+		cleanParameter(&c.Name),
+		c.Type.Clean(),
+	)
+}
+
+// NamedType is a single column in a
+// RETURN TABLE(...) statement in a procedure.
+type NamedType struct {
+	// Name is the name of the column.
+	Name string `json:"name"`
+	// Type is the type of the column.
+	Type *DataType `json:"type"`
+}
+
+func (p *NamedType) Clean() error {
+	return errors.Join(
+		cleanIdent(&p.Name),
+		p.Type.Clean(),
+	)
+}
+
+// DataType is a data type.
+// It includes both built-in types and user-defined types.
+type DataType struct {
+	// Name is the name of the type.
+	Name string `json:"name"`
+	// IsArray is true if the type is an array.
+	IsArray bool `json:"isArray"`
+}
+
+// String returns the string representation of the type.
+func (c *DataType) String() string {
+	if c.IsArray {
+		return c.Name + "[]"
+	}
+	return c.Name
+}
+
+// PGString returns the string representation of the type in Postgres.
+func (c *DataType) PGString() (string, error) {
+	var scalar string
+	switch strings.ToLower(c.Name) {
+	case "int":
+		scalar = "INT8"
+	case "text":
+		scalar = "TEXT"
+	case "bool":
+		scalar = "BOOL"
+	case "blob":
+		scalar = "BYTEA"
+	case "uuid":
+		scalar = "UUID"
+	case "null":
+		return "", fmt.Errorf("cannot have null column type")
+	case "unknown":
+		return "", fmt.Errorf("cannot have unknown column type")
+	default:
+		return "", fmt.Errorf("unknown column type: %s", c.Name)
+	}
+
+	if c.IsArray {
+		return scalar + "[]", nil
+	}
+
+	return scalar, nil
+}
+
+func (c *DataType) Clean() error {
+	c.Name = strings.ToLower(c.Name)
+
+	if c.Name != "int" &&
+		c.Name != "text" &&
+		c.Name != "bool" &&
+		c.Name != "blob" &&
+		c.Name != "uuid" &&
+		c.Name != "null" {
+		return fmt.Errorf("unknown column type: %s", c.Name)
+	}
+
+	return nil
+}
+
+// Copy returns a copy of the type.
+func (c *DataType) Copy() *DataType {
+	return &DataType{
+		Name:    c.Name,
+		IsArray: c.IsArray,
+	}
+}
+
+// Equals returns true if the type is equal to the other type.
+// If either type is Unknown, it will return true.
+func (c *DataType) Equals(other *DataType) bool {
+	if c.Name == "unknown" || other.Name == "unknown" {
+		return true
+	}
+	return strings.EqualFold(c.Name, other.Name) && c.IsArray == other.IsArray
+}
+
+var (
+	IntType = &DataType{
+		Name: "int",
+	}
+	TextType = &DataType{
+		Name: "text",
+	}
+	BoolType = &DataType{
+		Name: "bool",
+	}
+	BlobType = &DataType{
+		Name: "blob",
```

Comment:
Each of this string literals like `"blob"` are used three times in this file.   I think it's best to have a const block at the top defining them.